lib/segment/src/spaces/simple.rs

```diff
@@ -74,23 +60,111 @@ impl Metric for CosineMetric {
     }
 
     fn similarity(&self, v1: &[VectorElementType], v2: &[VectorElementType]) -> ScoreType {
-        v1.iter().zip(v2).map(|(a, b)| a * b).sum()
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            if is_x86_feature_detected!("avx2") {
+                return unsafe { dot_similarity_avx2(v1, v2) };
+            }
+        }
+        dot_similarity(v1, v2)
     }
 
-    fn blas_similarity(
-        &self,
-        v1: &Array1<VectorElementType>,
-        v2: &Array1<VectorElementType>,
-    ) -> ScoreType {
-        v1.dot(v2)
+    fn preprocess(&self, vector: &[VectorElementType]) -> Option<Vec<VectorElementType>> {
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            if is_x86_feature_detected!("avx2") {
+                return Some(unsafe { cosine_preprocess_avx2(vector) });
+            }
+        }
+        Some(cosine_preprocess(vector))
     }
+}
 
-    fn preprocess(&self, vector: &[VectorElementType]) -> Option<Vec<VectorElementType>> {
-        let mut length: f32 = vector.iter().map(|x| x * x).sum();
-        length = length.sqrt();
-        let norm_vector = vector.iter().map(|x| x / length).collect();
-        Some(norm_vector)
+fn euclid_similarity(v1: &[VectorElementType], v2: &[VectorElementType]) -> ScoreType {
+    let s: ScoreType = v1
+        .iter()
+        .copied()
+        .zip(v2.iter().copied())
+        .map(|(a, b)| (a - b).powi(2))
+        .sum();
+    -s.sqrt()
+}
+
+fn cosine_preprocess(vector: &[VectorElementType]) -> Vec<VectorElementType> {
+    let mut length: f32 = vector.iter().map(|x| x * x).sum();
+    length = length.sqrt();
+    vector.iter().map(|x| x / length).collect()
+}
+
+fn dot_similarity(v1: &[VectorElementType], v2: &[VectorElementType]) -> ScoreType {
+    v1.iter().zip(v2).map(|(a, b)| a * b).sum()
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "avx2")]
+unsafe fn hsum256_ps_avx(x: __m256) -> f32 {
```

Comment:
That looks pretty complicated! What is the source of this example?

Reply:
This function calculates the horizontal sum of all __m256 elements. I adopted an example from this conversation:
https://www.py4u.net/discuss/73145
Suggested implementation from the issue (https://github.com/zhao-lang/redis_hnsw/blob/master/src/hnsw/metrics.rs) has complicated __m256 sum too.